voxblox/src/integrator/tsdf_integrator.cc

```diff
@@ -0,0 +1,567 @@
+
+  timing::Timer integrate_timer("integrate");
+
+  ThreadSafeIndex index_getter(points_C.size(), config_.integrator_threads);
+
+  std::vector<std::thread> integration_threads;
+  for (size_t i = 0; i < config_.integrator_threads; ++i) {
+    integration_threads.emplace_back(&SimpleTsdfIntegrator::integrateFunction,
+                                     this, T_G_C, points_C, colors,
+                                     &index_getter, &(temp_voxel_storage[i]));
+  }
+
+  for (std::thread& thread : integration_threads) {
+    thread.join();
+  }
+  integrate_timer.Stop();
+
+  timing::Timer insertion_timer("inserting_missed_voxels");
+  for (const VoxelMap& temp_voxels : temp_voxel_storage) {
+    updateLayerWithStoredVoxels(temp_voxels);
+  }
+  insertion_timer.Stop();
+}
+
+void SimpleTsdfIntegrator::integrateFunction(const Transformation& T_G_C,
+                                             const Pointcloud& points_C,
+                                             const Colors& colors,
+                                             ThreadSafeIndex* index_getter,
+                                             VoxelMap* temp_voxel_storage) {
+  DCHECK(index_getter != nullptr);
+  DCHECK(temp_voxel_storage != nullptr);
+
+  size_t point_idx;
+  while (index_getter->getNextIndex(&point_idx)) {
+    const Point& point_C = points_C[point_idx];
+    const Color& color = colors[point_idx];
+    bool is_clearing;
+    if (!isPointValid(point_C, &is_clearing)) {
+      continue;
+    }
+
+    const Point origin = T_G_C.getPosition();
+    const Point point_G = T_G_C * point_C;
+
+    RayCaster ray_caster(origin, point_G, is_clearing,
+                         config_.voxel_carving_enabled,
+                         config_.max_ray_length_m, voxel_size_inv_,
+                         config_.default_truncation_distance);
+
+    Block<TsdfVoxel>::Ptr block = nullptr;
+    BlockIndex block_idx;
+    VoxelIndex global_voxel_idx;
+    while (ray_caster.nextRayIndex(&global_voxel_idx)) {
+      TsdfVoxel* voxel = findOrTempAllocateVoxelPtr(
+          global_voxel_idx, &block, &block_idx, temp_voxel_storage);
+
+      const float weight = getVoxelWeight(point_C);
+      const Point voxel_center_G =
+          getCenterPointFromGridIndex(global_voxel_idx, voxel_size_);
+
+      updateTsdfVoxel(origin, point_G, voxel_center_G, color, weight, voxel);
+    }
+  }
+}
+
+void MergedTsdfIntegrator::integratePointCloud(const Transformation& T_G_C,
+                                               const Pointcloud& points_C,
+                                               const Colors& colors) {
+  timing::Timer integrate_timer("integrate");
+
+  // Pre-compute a list of unique voxels to end on.
+  // Create a hashmap: VOXEL INDEX -> index in original cloud.
+  BlockHashMapType<std::vector<size_t>>::type voxel_map;
+  // This is a hash map (same as above) to all the indices that need to be
+  // cleared.
+  BlockHashMapType<std::vector<size_t>>::type clear_map;
```

Comment:
I guess it does make sense to rename BlockHasMapType, because here it clearly referes to voxel indices, right?